File: packages/synth/wasm/src/runtime_dispatch/envelopes.rs
use crate::runtime_dispatch::RenderContext;
use crate::runtime_envelopes::{render_adsr, render_ar, render_exponential};
use crate::runtime_node::RuntimeNode;
use crate::runtime_state::NodeBuffer;

#[inline(always)]
pub(super) fn render_envelope_node(
    node: &mut RuntimeNode,
    ctx: &RenderContext<'_>,
    out: &mut NodeBuffer,
) -> Option<()> {
    match node {
        RuntimeNode::Adsr {
            input,
            attack,
            decay,
            sustain,
            release,
            ..
        } => render_adsr(
            *input,
            *attack,
            *decay,
            *sustain,
            *release,
            ctx.access,
            ctx.prior,
            ctx.params,
            out,
            ctx.frames,
            ctx.sample_rate,
            ctx.frame_start,
            ctx.gate_sec,
            ctx.velocity,
        ),
        RuntimeNode::Ar {
            input,
            attack,
            release,
            ..
        } => render_ar(
            *input,
            *attack,
            *release,
            ctx.access,
            ctx.prior,
            ctx.params,
            out,
            ctx.frames,
            ctx.sample_rate,
            ctx.frame_start,
            ctx.gate_sec,
            ctx.velocity,
        ),
        RuntimeNode::Exponential { input, tau, .. } => render_exponential(
            *input,
            *tau,
            ctx.access,
            ctx.prior,
            ctx.params,
            out,
            ctx.frames,
            ctx.sample_rate,
            ctx.frame_start,
        ),
        _ => unreachable!("non-envelope node dispatched to envelope renderer"),
    }
}

File: packages/synth/src/nodes/envelopes.ts
import type { Transform } from '../types'
import { common } from './shared'

type EnvelopeFn = ((opts: {
  attack: number
  decay: number
  sustain: number
  release: number
}) => Transform) & {
  ar(opts: { attack: number; release: number }): Transform
  exponential(opts: { tau: number }): Transform
}

const envelopeBase =
  (opts: { attack: number; decay: number; sustain: number; release: number }): Transform =>
  (node) => ({
    kind: 'adsr',
    input: node,
    attack: opts.attack,
    decay: opts.decay,
    sustain: opts.sustain,
    release: opts.release,
    ...common(node.out),
  })

export const envelope = Object.assign(envelopeBase, {
  ar:
    (opts: { attack: number; release: number }): Transform =>
    (node) => ({
      kind: 'ar',
      input: node,
      attack: opts.attack,
      release: opts.release,
      ...common(node.out),
    }),
  exponential:
    (opts: { tau: number }): Transform =>
    (node) => ({ kind: 'exponential', input: node, tau: opts.tau, ...common(node.out) }),
}) as EnvelopeFn
